### PR TITLE
RGD code cleanup

### PR DESCRIPTION
**Pull request by ptosco**
Ahead of implementing a new feature I had to remind myself how RGD works as it had been a while since I looked at that code.
While doing that I noticed a number of things which were not clear. which looked like dead code or which could be implemented more efficiently, and I thought of committing that as a cleanup PR.

- made an effort to give more meaningful names to variables (e.g., renamed most instances of attachment (point) to avoid ambiguity as attachment may be interpreted as either the R-group atom or its neighbor atom on the core, which are two different things)
- replaced the old school `removeAtom()` method with `begin/commitBatchEdit()`
- added `std::move` and `std::make_move_iterator` where relevant to avoid potential unintended copying
- replaced instances of `container.size() == 0` and `!container.size()` with `container.empty()` for better clarity
- replaced `std::map::find()` with `std::map::at()` where the key was not needed
- replaced expensive `std::find_if` with more efficient alternative
- added some missing `const` keywords and added references to avoid copying where appropriate
- replaced for loops with modern implicit looping alternatives where convenient
- avoid calling `MolToSmiles` when `VERBOSE` is not defined as the result is anyway not used
- removed "oops, exponential is a pain" code snippet as 1. it is almost certainly never executed 2. it is not tested 3. I do not think it is correct
- removed check for `data->matches.size() > 1` as I do not believe it is correct
- Use `std::unique_ptr::reset` instead of defining a new `std::unique_ptr` and moving it to the original one

### CODE REVIEW

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp [hunk 1]
```diff
@@ -17,16 +17,16 @@
...
-void RGroupData::add(boost::shared_ptr<ROMol> newMol,
+void RGroupData::add(const ROMOL_SPTR &newMol,
                      const std::vector<int> &rlabel_attachments) {
-  // some fragments can be add multiple times if they are cyclic
-  for (auto &mol : mols) {
-    if (newMol.get() == mol.get()) {
-      return;
-    }
+  // some fragments can be added multiple times if they are cyclic
+  if (std::any_of(mols.begin(), mols.end(), [&newMol](const auto &mol) {
```
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupData.cpp:23]: I find the previous code a lot more clear than this.

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp [hunk 2]
```diff
@@ -94,26 +94,16 @@ std::string RGroupData::toString() const {
...
 
 void RGroupData::computeIsHydrogen() {  // is the rgroup all Hs
-  for (const auto &mol : mols) {
-    if (!isMolHydrogen(*mol)) {
-      is_hydrogen = false;
-      return;
-    }
-  }
-  is_hydrogen = true;
+  is_hydrogen = std::all_of(mols.begin(), mols.end(), [this](const auto &mol) {
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:97]: (previous code is clearer)

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp [hunk 3]
```diff
@@ -94,26 +94,16 @@ std::string RGroupData::toString() const {
...
-    if (atom->getAtomicNum() > 1) {
-      return false;
-    } else if (atom->getAtomicNum() == 0 && !atom->hasProp(SIDECHAIN_RLABELS)) {
-      return false;
-    }
-  }
-  return true;
+bool RGroupData::isMolHydrogen(const ROMol &mol) const {
+  auto atoms = mol.atoms();
+  return std::all_of(atoms.begin(), atoms.end(), [](const auto &atom) {
```
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupData.cpp:104]: Interestingly, I find this more clear than the previous...

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp [hunk 4]
```diff
@@ -178,60 +180,53 @@ int RGroupDecomposition::getMatchingCoreInternal(
     size_t i = 0;
     for (const auto &mv : tmatches) {
       bool passes_filter = data->params.onlyMatchAtRGroups;
-      boost::dynamic_bitset<> target_match_indices(mol.getNumAtoms());
+      std::vector<std::vector<int>> targetToCoreIndices(mol.getNumAtoms());
```
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:183]: This data structure is used quite a bit now, we should document what it is doing.  It appears to be mapping

atomIndex -> vector of possible matches?

**Comment by bp-kelley** [comment]
@ptosco thanks for the cleanup.  At some point I need to just sit down and write the algorithm doc for this.  Like most things it started out pretty simple but then devolved into madness.

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:23]: For me, the main advantage of `std::all_of` and `std::any_of` is that they allow to tell at a glance whether all, any or none (for the `!std::any_of` case) of the container items need to fulfil the condition described in the lambda without having to drill down into the logic of `for` loops.

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:97]: Ditto.

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp (hunk 3 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:104]: You see? You are already hooked :-)

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:23]: We can meet halfway here with I suppose and use the shared_ptrs == operator.  This reads better than the original.

```
if (std::any_of(mols.begin(), mols.end(),
		        [&newMol](const auto &mol) {
		          return newMol == mol;})
    ) {
  return;
 }
```

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp (hunk 4 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:194]: Good point, I added a detailed explanation as it is not obvious.

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp [hunk 5]
```diff
@@ -178,60 +180,53 @@ int RGroupDecomposition::getMatchingCoreInternal(
...
       }
 
       if (passes_filter) {
-        tmatches_filtered.push_back(mv);
+        tmatches_filtered.push_back(std::move(mv));
       }
       ++i;
     }
     if (!data->params.onlyMatchAtRGroups) {
+      CHECK_INVARIANT(!tmatches_heavy_nbrs.empty(), "");
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp]: Done.

File: Code/GraphMol/RGroupDecomposition/RGroupData.cpp (hunk 1 above)
Comment (ptosco) [reply Code/GraphMol/RGroupDecomposition/RGroupData.cpp:23]: Ha, you added your suggestion while I was committing the revisions based on your previous remarks. I have just committed this last suggestion.

**Comment by bp-kelley** [comment]
The windows failures are real, but a bit confusing.  Are we setting WARNING logs to be errors here?  Or is this a real exception somewhere.

**Comment by ptosco** [comment]
@bp-kelley I believe the warning is printed because the test crashes and hence `stderr` is dumped to the CI log. I'll take a look at what's wrong with my last commit leter today.

**Comment by ptosco** [comment]
@bp-kelley Fixed.

**Review by bp-kelley — APPROVED** [review]

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp [hunk 6]
```diff
@@ -154,20 +154,22 @@ int RGroupDecomposition::getMatchingCoreInternal(
...
-        "matchesWithDummy "; for (const auto &matchWithDummy : matchesWithDummy)
+        "matchesIncludingRGroups "; for (const auto &matchWithDummy : matchesIncludingRGroups)
         { for (const auto &pair : matchWithDummy) std::cerr << "(" << pair.first
         <<"," << pair.second << "),"; std::cerr << " /// ";
         }
         std::cerr << std::endl;
         */
-        tmatches.insert(tmatches.end(), matchesWithDummy.cbegin(),
-                        matchesWithDummy.cend());
+        tmatches.insert(tmatches.end(), std::make_move_iterator(matchesIncludingRGroups.cbegin()),
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:171]: Nice. Let's see if I remember this one.

**Review by greglandrum — APPROVED** [review]
LGTM

**Pull request merged**